app.js

```diff
@@ -47,16 +55,100 @@ io.on('connection', function(socket) {
 	socket.on('Message Request', function(data){
 
 		//if message is valid
-		if(data) {
+		if(data.msg) {
 
-			//display message to all clients in lobby including sender
-			io.sockets.emit('Display Message', {msg: data, user: socket.username});
+			//display message to all clients in room including sender
+			io.sockets["in"](data.room).emit('Display Message', {msg: data.msg, user: socket.username, room: data.room});
 		}
 	});
 
+	//When client creates room
+	socket.on('create room', function(data) {
+		data.room_name = data.room_name.trim();
+
+		//if room name is empty, do nothing
+		if(data.room_name == null) {
+			return;
+		}
+
+		var num_rooms = rooms.length
+
+		//check if room exists
+		for(var i = 0; i<num_rooms; ++i) {
+
+			//if room name is taken
+			if(rooms[i][0] == data.room_name) {
+				socket.emit('room exists', socket.username);
+				return;
+			}
+		}
+
+		//room not taken so insert into room array 
+		rooms.push([data.room_name, 0]);
+		io.sockets.emit('room created', data);
+		
+	});
+
+	//When user requests to join the room
+	socket.on('join room', function(room) {
+		socket.join(room);
+
+		var num_rooms = rooms.length
+
+		//update number of users in room
+		for(var i = 0; i < num_rooms; i++) {
+			if(room == rooms[i][0]) {
+				rooms[i][1]++;
+				break;
+			}
+		}
+
+		//notify other users in room that someone joined
+		io.sockets["in"](room).broadcast.emit('user join', {username: socket.username, room: room});
+	});
+
+	//When user requests to leave the room
+	socket.on('leave room', function(room) {
+		socket.leave(room);
+
+		var num_rooms = rooms.length;
+
+		//update number of users in room
+		for(var i = 0; i < num_rooms; i++) {
+			if(room == rooms[i][0]) {
+				rooms[i][1]--;
+
+				//if users become 0, destroy/delete the room
+				if(rooms[i][1] == 0) {
+					io.sockets.emit('destroy room', room);
+					rooms.splice(i, 1);
+				}
+				break;
+			}
+		}
+
+		//notify other users in room that someone left
+		io.sockets["in"](room).broadcast.emit('user left', {username: socket.username, room: room});
+	});
+
 	//When user disconnets remove user from users
 	socket.on('disconnect', function() {
 
+		var num_rooms = rooms.length;
+
+		//update number of users in rooms 
+		for(var i = 0; i < num_rooms; i++) {
+			if(socket.rooms.indexOf(rooms[i][0]) >= 0) {	
```

Comment:
Have you tested this?
Once disconnected I think socket.rooms would return an empty array.